### PR TITLE
tidb-server: dump heap profile automatically when heap size growing

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?

Problem Summary:
It is hard to debug why TiDB killed by oom in the production environments.

### What is changed and how it works?
Use a background worker to check the current heap size, dump heap profile and goroutines if needed. These profiles will be placed in the `path/of/slow-log/profile`.

The filename of the output file is `time-pid-size-type`.

图片

### Check List 

Tests 
- Manual test (add detailed scripts or steps below)

### Release note 
* dump heap profile automatically when heap size growing.

### CODE REVIEW

**Comment by jackysp** [comment]
PTAL @coocood

File: tidb-server/main.go [hunk 1]
```diff
@@ -730,3 +732,58 @@ func stringToList(repairString string) []string {
...
+	}()
+}
+
+func dumpProfile(path string, threshold uint64) error {
+	time := time.Now().Format("2006-01-02T15:04:05")
+	fname := func(t string) string {
+		return fmt.Sprintf("%s-%d-%dGiB-%s", time, os.Getpid(), threshold, t)
+	}
+
+	hf, err := os.Create(filepath.Join(path, fname("heap.pb.gz")))
```
Comment (coocood) [inline tidb-server/main.go:771]: The file is gzipped automatically?

File: tidb-server/main.go [hunk 2]
```diff
@@ -730,3 +732,58 @@ func stringToList(repairString string) []string {
...
+	if err != nil {
+		return err
+	}
+	defer hf.Close()
+	if err := pprof.Lookup("heap").WriteTo(hf, 0); err != nil {
```
Comment (coocood) [inline tidb-server/main.go:776]: Why not set debug = 1?

File: tidb-server/main.go [hunk 3]
```diff
@@ -730,3 +732,58 @@ func stringToList(repairString string) []string {
...
+			var mem runtime.MemStats
+			runtime.ReadMemStats(&mem)
+			if mem.HeapAlloc < nextThresholdGiB<<30 {
+				continue
+			}
+			currentGiB := mem.HeapAlloc >> 30
+			if err := dumpProfile(profileDir, currentGiB); err != nil {
+				log.Info("dump profile failed", zap.Error(err))
+			}
+			nextThresholdGiB = currentGiB + 1
```
Comment (coocood) [inline tidb-server/main.go:759]: Should we increase the step when currentGB is large?

File: tidb-server/main.go (hunk 1 above)
Comment (lysu) [inline tidb-server/main.go:771]: maybe we also need a cleanup mechanism to “remove old files“ or "keep N files" ?

**Pull request closed without merging**